Failed migration 0.11.1 -> 0.14.0-RC3

Turns out #2405 did not help. I tried upgrade from 0.11.1 -> 0.14.0-RC3 before, but it failed. New run fails too:

```
[2016-01-13 23:59:04,275] ERROR Failed to take over leadership (mesosphere.marathon.MarathonSchedulerService$$EnhancerByGuice$$348337d4:pool-1-thread-2)
mesosphere.marathon.MigrationFailedException: while migrating storage to 0.13
	at mesosphere.marathon.state.Migration$$anonfun$migrations$4$$anonfun$apply$2.applyOrElse(Migration.scala:61) ~[marathon-assembly-0.14.0-RC3.jar:0.14.0-RC3]
	at mesosphere.marathon.state.Migration$$anonfun$migrations$4$$anonfun$apply$2.applyOrElse(Migration.scala:60) ~[marathon-assembly-0.14.0-RC3.jar:0.14.0-RC3]
	at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:36) ~[marathon-assembly-0.14.0-RC3.jar:0.14.0-RC3]
	at scala.util.Failure$$anonfun$recover$1.apply(Try.scala:216) ~[marathon-assembly-0.14.0-RC3.jar:0.14.0-RC3]
	at scala.util.Try$.apply(Try.scala:192) ~[marathon-assembly-0.14.0-RC3.jar:0.14.0-RC3]
	at scala.util.Failure.recover(Try.scala:216) ~[marathon-assembly-0.14.0-RC3.jar:0.14.0-RC3]
	at scala.concurrent.Future$$anonfun$recover$1.apply(Future.scala:324) ~[marathon-assembly-0.14.0-RC3.jar:0.14.0-RC3]
	at scala.concurrent.Future$$anonfun$recover$1.apply(Future.scala:324) ~[marathon-assembly-0.14.0-RC3.jar:0.14.0-RC3]
	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:32) ~[marathon-assembly-0.14.0-RC3.jar:0.14.0-RC3]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_66-internal]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_66-internal]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_66-internal]
Caused by: mesosphere.marathon.StoreCommandFailedException: Can not create entity framework:id
	at mesosphere.util.state.zk.ZKStore$$anonfun$exceptionTransform$1.applyOrElse(ZKStore.scala:84) ~[marathon-assembly-0.14.0-RC3.jar:0.14.0-RC3]
	at mesosphere.util.state.zk.ZKStore$$anonfun$exceptionTransform$1.applyOrElse(ZKStore.scala:83) ~[marathon-assembly-0.14.0-RC3.jar:0.14.0-RC3]
	... 10 common frames omitted
Caused by: org.apache.zookeeper.KeeperException$NodeExistsException: KeeperErrorCode = NodeExists for /marathons/scan/state/framework:id
	at org.apache.zookeeper.KeeperException.create(KeeperException.java:119) ~[marathon-assembly-0.14.0-RC3.jar:0.14.0-RC3]
	at org.apache.zookeeper.KeeperException.create(KeeperException.java:51) ~[marathon-assembly-0.14.0-RC3.jar:0.14.0-RC3]
	at com.twitter.zk.AsyncCallbackPromise$class.process(AsyncCallbackPromise.scala:18) ~[marathon-assembly-0.14.0-RC3.jar:0.14.0-RC3]
	at com.twitter.zk.StringCallbackPromise.process(AsyncCallbackPromise.scala:23) ~[marathon-assembly-0.14.0-RC3.jar:0.14.0-RC3]
	at com.twitter.zk.StringCallbackPromise.processResult(AsyncCallbackPromise.scala:25) ~[marathon-assembly-0.14.0-RC3.jar:0.14.0-RC3]
	at org.apache.zookeeper.ClientCnxn$EventThread.processEvent(ClientCnxn.java:605) ~[marathon-assembly-0.14.0-RC3.jar:0.14.0-RC3]
	at org.apache.zookeeper.ClientCnxn$EventThread.run(ClientCnxn.java:498) ~[marathon-assembly-0.14.0-RC3.jar:0.14.0-RC3]
```

`frameworkId` znode exists, but not `framework:id`.

This only happened to one of 19 marathons we have and it can be easily fixed by redeploying, but still.

--- a/src/main/scala/mesosphere/marathon/state/Migration.scala
+++ b/src/main/scala/mesosphere/marathon/state/Migration.scala
@@ -364,9 +364,20 @@ class MigrationTo0_13(taskRepository: TaskRepository, store: PersistentStore) {
         _ <- store.delete(oldName)
       } yield ()
     }
-    store.load(oldName).flatMap {
-      case Some(key) => moveKey(key.bytes)
-      case None      => Future.successful(())
+
+    store.load(newName).flatMap {
+      case Some(_) =>
+        log.info("framework:id already exists, no need to migrate")
+        Future.successful(())
+      case None =>
+        store.load(oldName).flatMap {
+          case None =>
+            log.info("no frameworkId stored, no need to migrate")
+            Future.successful(())
+          case Some(entity) =>
+            log.info("migrating frameworkId -> framework:id")
+            moveKey(entity.bytes)
+        }
     }
   }
 

--- a/src/test/scala/mesosphere/marathon/state/MigrationTo0_13Test.scala
+++ b/src/test/scala/mesosphere/marathon/state/MigrationTo0_13Test.scala
@@ -10,6 +10,7 @@ import mesosphere.marathon.MarathonSpec
 import mesosphere.marathon.Protos.MarathonTask
 import mesosphere.marathon.metrics.Metrics
 import mesosphere.marathon.state.PathId.StringPathId
+import mesosphere.util.state.FrameworkId
 import mesosphere.util.state.memory.InMemoryStore
 import org.scalatest.{ GivenWhenThen, Matchers }
 
@@ -92,6 +93,62 @@ class MigrationTo0_13Test extends MarathonSpec with GivenWhenThen with Matchers 
     taskKeys2 should not contain f.legacyStoreKey(appId, task2.getId)
   }
 
+  test("migrating frameworkId to framework:id") {
+    import FrameworkIdValues._
+    val f = new Fixture
+    val frameworkId = FrameworkId("myFramework")
+
+    Given("a frameworkId under the old key")
+    f.frameworkIdStore.store(oldName, frameworkId).futureValue
+    f.frameworkIdStore.names().futureValue should contain (oldName)
+    f.frameworkIdStore.fetch(oldName).futureValue should be (Some(frameworkId))
+
+    When("we run the migration")
+    f.migration.renameFrameworkId().futureValue
+
+    Then("The old key should be deleted")
+    val namesAfterMigration = f.frameworkIdStore.names().futureValue
+    namesAfterMigration should not contain oldName
+
+    And("The new key should be present and contain the correct value")
+    namesAfterMigration should contain (newName)
+    f.frameworkIdStore.fetch(newName).futureValue should be (Some(frameworkId))
+  }
+
+  test("migrating frameworkId does nothing for non-existent node") {
+    val f = new Fixture
+
+    Given("a non-existing frameworkId")
+    f.frameworkIdStore.names().futureValue should be (empty)
+
+    When("we run the migration")
+    f.migration.renameFrameworkId().futureValue
+
+    Then("Nothing should have happened")
+    f.frameworkIdStore.names().futureValue should be (empty)
+  }
+
+  test("migrating frameworkId is skipped if framework:id already exists") {
+    import FrameworkIdValues._
+    val f = new Fixture
+    val frameworkId = FrameworkId("myFramework")
+
+    Given("An existing framework:id")
+    f.frameworkIdStore.store(newName, frameworkId).futureValue
+    val names = f.frameworkIdStore.names().futureValue
+    names.size should be (1)
+    names should contain (newName)
+
+    When("we run the migration")
+    f.migration.renameFrameworkId().futureValue
+
+    Then("Nothing should have changed")
+    val newNames = f.frameworkIdStore.names().futureValue
+    newNames.size should be (1)
+    newNames should contain (newName)
+    f.frameworkIdStore.fetch(newName).futureValue should be (Some(frameworkId))
+  }
+
   class Fixture {
     lazy val uuidGenerator = Generators.timeBasedGenerator(EthernetAddress.fromInterface())
     lazy val state = new InMemoryStore
@@ -106,9 +163,21 @@ class MigrationTo0_13Test extends MarathonSpec with GivenWhenThen with Matchers 
       val metrics = new Metrics(new MetricRegistry)
       new TaskRepository(entityStore, metrics)
     }
+    lazy val frameworkIdStore = new MarathonStore[FrameworkId](
+      store = state,
+      metrics = metrics,
+      newState = () => new FrameworkId(UUID.randomUUID().toString),
+      prefix = "" // don't set the prefix so we don't have to use PersistentStore for testing
+    )
+
     lazy val migration = new MigrationTo0_13(taskRepo, state)
 
     def legacyStoreKey(appId: PathId, taskId: String): String = appId.safePath + ":" + taskId
+  }
+
+  object FrameworkIdValues {
+    val oldName = "frameworkId"
+    val newName = "framework:id"
   }
 
 }

